db/utils.js

```diff
@@ -49,6 +49,24 @@ const DBUtils = {
     }
   },
 
+  async _getSha1EntriesFromPrefixAndDo(sha1Prefix, aFoundCallback, aNotFoundCallback) {
+    // Only accept 6-character hash prefixes so requests:
+    // 1. can't get more than the intended hash range results FROM us
+    // 2. can't reveal a more specific hash query TO us
+    if (sha1Prefix.length !== 6) {
+      return await aNotFoundCallback();
+    }
+    const existingEntries = await EmailHash.query().where("sha1", "like", sha1Prefix + "%").eager("breaches");
+
+    if (existingEntries.length && aFoundCallback) {
+      return await aFoundCallback(existingEntries);
+    }
+
+    if (aNotFoundCallback) {
```

Comment:
This will be called even if we got existing entries, if the caller of the function didn't supply `aFoundCallback`. We need to either return early before this check or add a `!existingEntries.length` to the condition - I prefer the former.

Reply:
I chose the latter approach so I didn't accidentally change any caller's behavior by assuming when I should return early.

I'd honestly like to get rid of these `get...AndDo` functions completely. I find `andDo(... [callback, ...])` functions hard to fit into my brain. 😐